### PR TITLE
[android] Add @shopify/react-native-skia vendoring support for Expo Go 

### PR DESCRIPTION
**Pull request by Kudo**
# Why

like #17704 but for android
close ENG-4889

# How

- introduce vendor modules autolinking for expo go.
previously, reanimated is the only vendor module having cxx code, we had a fixed versioning script for reanimated. because react-native-skia also has cxx code, i'm trying to keep the versioning more modularized and like what we did for ios versioning. ideally, a vendor module should have a dedicated gradle project. that's why i introduced a vendor module autolinking here.
- vendoring tool: add react-native-skia support
- versioning tool: trying to add a minimum support for new style versioning like ios.
currently, after transforming the code, i still prebuild the *.so to `jniLibs`. in the future when we commit `android/versioned-react-native/ABI46_0_0` to git, we can always build the versioned *.so from source. 

# Test Plan

- android unversioned expo go + NCL skia
- android versioned expo go + NCL skia

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by Kudo** [comment]
the android client is intentionally breaking because i didn't adding skia to vendor module in this pr. after the pr getting reviewed, i'll wait to merge it and add skia vendor code together.

**Comment by Kudo** [comment]
cc @wkozyra95 because this pr have some versioning tool changes.

File: tools/src/versioning/android/versionVendoredModules.ts [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        replaceWith: `$1${prefix}/`,
+      },
+      {
+        paths: 'build.gradle',
+        find: `implementation 'com.facebook.react:react-native:+'`,
+        replaceWith:
+          `implementation 'host.exp:reactandroid-${prefix}:1.0.0'` +
+          '\n' +
+          // Adding some compile time common dependencies where the versioned react-native AAR doesn't expose
+          `    compileOnly 'com.facebook.fbjni:fbjni:0.2.2'\n` +
```
Comment (wkozyra95) [inline tools/src/versioning/android/versionVendoredModules.ts:179]: hardcoding versions in versioning script? wouldn't this require manual changes every time we update rn?

File: tools/src/versioning/android/versionVendoredModules.ts [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+
+    await maybePrebuildSharedLibsAsync(name, sdkNumber);
+  }
+}
+
+/**
+ * Prebuild shared libraries to jniLibs and cleanup CMakeLists.txt
+ */
+async function maybePrebuildSharedLibsAsync(module: string, sdkNumber: number) {
+  const moduleRootDir = path.join(ANDROID_DIR, 'vendored', `sdk${sdkNumber}`, module, 'android');
```
Comment (wkozyra95) [inline tools/src/versioning/android/versionVendoredModules.ts:64]: why are you adding new type of identifier for SDK here, everything on android is using `abix_0_0`

File: tools/src/versioning/android/versionVendoredModules.ts [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+
+/**
+ * Generates base transforms to apply for all vendored modules.
+ */
+function baseTransformsFactory(prefix: string): Required<FileTransforms> {
+  return {
+    path: [
+      {
+        // For package renaming, src/main/java/* -> src/main/java/abiN/*
+        find: /\/java\//,
```
Comment (wkozyra95) [inline tools/src/versioning/android/versionVendoredModules.ts:147]: maybe match entire `./src/main/java` ?

**Review by wkozyra95 — APPROVED** [review]

File: tools/src/vendoring/config/expoGoConfig.ts [hunk 4]
```diff
@@ -307,6 +310,41 @@ const config: VendoringTargetConfig = {
...
+            'src',
+            'vendoring',
+            'config',
+            'react-native-skia.patch'
+          );
+          const patchContent = await fs.readFile(
+            patchFile,
+
+            'utf8'
+          );
```
Comment (tsapeta) [inline tools/src/vendoring/config/expoGoConfig.ts:333]:
```suggestion
          const patchContent = await fs.readFile(patchFile, 'utf8');
```

File: tools/src/Patch.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import spawnAsync from '@expo/spawn-async';
+
+export async function runPatchAsync(options: {
```
Comment (tsapeta) [inline tools/src/Patch.ts:3]: As it's just one function, maybe it makes sense to move it to `Utils.ts`? Also, I think "apply" would be better than "run" 😉

**Comment by Kudo** [comment]
i'm going to merge this. if there're any further comments, please let me know and i'll address in separated pr. thanks!

**Pull request merged**